tasks/messageexecutions/message.go

```diff
@@ -0,0 +1,194 @@
+	"go.opentelemetry.io/otel/label"
+	"golang.org/x/xerrors"
+)
+
+var log = logging.Logger("messageexecutions")
+
+func NewTask() *Task {
+	return &Task{}
+}
+
+type Task struct {
+}
+
+func (p *Task) Close() error {
+	return nil
+}
+
+func (p *Task) ProcessMessageExecutions(ctx context.Context, store adt.Store, ts *types.TipSet, pts *types.TipSet, mex []*lens.MessageExecution) (model.Persistable, *visormodel.ProcessingReport, error) {
+	ctx, span := global.Tracer("").Start(ctx, "ProcessMessageExecutions")
+	if span.IsRecording() {
+		span.SetAttributes(label.String("tipset", ts.String()), label.Int64("height", int64(ts.Height())))
+	}
+	defer span.End()
+
+	report := &visormodel.ProcessingReport{
+		Height:    int64(pts.Height()),
+		StateRoot: pts.ParentState().String(),
+	}
+
+	var (
+		internalResult       = make(messagemodel.InternalMessageList, 0, len(mex))
+		internalParsedResult = make(messagemodel.InternalParsedMessageList, 0, len(mex))
+		errorsDetected       = make([]*messages.MessageError, 0) // we don't know the cap since mex is recursive in nature.
+	)
+
+	for _, m := range mex {
+		select {
+		case <-ctx.Done():
+			return nil, nil, xerrors.Errorf("context done: %w", ctx.Err())
+		default:
+		}
+
+		// we don't yet record implicit messages in the other message task, record them here.
+		if m.Implicit {
+			toName, toFamily, err := util.ActorNameAndFamilyFromCode(m.ToActorCode)
+			if err != nil {
+				errorsDetected = append(errorsDetected, &messages.MessageError{
+					Cid:   m.Cid,
+					Error: xerrors.Errorf("failed get message to actor name and family: %w", err).Error(),
+				})
+			}
+			internalResult = append(internalResult, &messagemodel.InternalMessage{
+				Height:        int64(m.Height),
+				Cid:           m.Cid.String(),
+				SourceMessage: "", // there is no source for implicit messages, they include cron tick and reward messages only
+				StateRoot:     m.StateRoot.String(),
+				From:          m.Message.From.String(),
+				To:            m.Message.To.String(),
+				ActorName:     toName,
+				ActorFamily:   toFamily,
+				Value:         m.Message.Value.String(),
+				Method:        uint64(m.Message.Method),
+				ExitCode:      int64(m.Ret.ExitCode),
+				GasUsed:       m.Ret.GasUsed,
+			})
+			method, params, err := util.MethodAndParamsForMessage(m.Message, m.ToActorCode)
+			if err != nil {
+				errorsDetected = append(errorsDetected, &messages.MessageError{
+					Cid:   m.Cid,
+					Error: xerrors.Errorf("failed parse method and params for message: %w", err).Error(),
+				})
+			}
+			internalParsedResult = append(internalParsedResult, &messagemodel.InternalParsedMessage{
+				Height: int64(m.Height),
+				Cid:    m.Cid.String(),
+				From:   m.Message.From.String(),
+				To:     m.Message.To.String(),
+				Value:  m.Message.Value.String(),
+				Method: method,
+				Params: params,
+			})
+		}
+
+		// TODO(frrist): this code is commented out as it collects all internal message sent through the VM.
```

Comment:
issue for this here: https://github.com/filecoin-project/sentinel-visor/issues/602